### PR TITLE
MODSOURMAN-541 Update existing CLI endpoint GET /mapping-rules to support MARC Holdi…

### PR DESCRIPTION
**Pull request by Serhii-Zdorenko**
## Purpose
Update existing CLI endpoint GET /mapping-rules to support MARC Holdings default rules

## Approach
modify existing GET /mapping-rules endpoint to GET /mapping-rules/marc-holdings
An endpoint should support URI path parameters:
marc-holdings
marc-bib
and return corresponding rules

## Learning
https://issues.folio.org/browse/MODSOURMAN-541

### CODE REVIEW

File: mod-source-record-manager-server/src/main/java/org/folio/services/MappingRuleService.java [hunk 1]
```diff
@@ -14,17 +15,18 @@
    * Returns rules in JsonObject
    *
    * @param tenantId tenant
+   * @param recordType type of rules (MARC_BIB or MARK_HOLDING)
    * @return optional with rules in JsonObject
    */
-  Future<Optional<JsonObject>> get(String tenantId);
+  Future<Optional<JsonObject>> get(String tenantId, String recordType);
```
Comment (psmagin) [inline mod-source-record-manager-server/src/main/java/org/folio/services/MappingRuleService.java:21]: Will be better to have `tenantId` as last parameter. Please check this in all updated methods.

File: mod-source-record-manager-server/src/main/java/org/folio/services/MappingRuleServiceImpl.java [hunk 2]
```diff
@@ -33,18 +35,24 @@ public MappingRuleServiceImpl(MappingRuleDao mappingRuleDao, MappingRuleCache ma
...
+    }
   }
 
   @Override
-  public Future<Void> saveDefaultRules(String tenantId) {
+  public Future<Void> saveDefaultRules(String tenantId, Record.RecordType recordType) {
     Promise<Void> promise = Promise.promise();
-    Optional<String> optionalRules = readResourceFromPath(DEFAULT_RULES_PATH);
+    Optional<String> optionalRules = Optional.empty();
+    if (recordType == Record.RecordType.MARC_BIB) optionalRules = readResourceFromPath(DEFAULT_BIB_RULES_PATH);
```
Comment (psmagin) [inline mod-source-record-manager-server/src/main/java/org/folio/services/MappingRuleServiceImpl.java:50]: Please use curly braces in if

File: mod-source-record-manager-server/src/main/java/org/folio/services/MappingRuleServiceImpl.java [hunk 3]
```diff
@@ -33,18 +35,24 @@ public MappingRuleServiceImpl(MappingRuleDao mappingRuleDao, MappingRuleCache ma
...
+    if (recordType == Record.RecordType.MARC_HOLDING) optionalRules = readResourceFromPath(DEFAULT_HOLDING_RULES_PATH);
```
Comment (psmagin) [inline mod-source-record-manager-server/src/main/java/org/folio/services/MappingRuleServiceImpl.java:51]:
```suggestion
    else if (recordType == Record.RecordType.MARC_HOLDING) optionalRules = readResourceFromPath(DEFAULT_HOLDING_RULES_PATH);
```

File: mod-source-record-manager-server/src/main/java/org/folio/services/ParsedRecordServiceImpl.java [hunk 4]
```diff
@@ -30,6 +15,19 @@
...
+import javax.ws.rs.NotFoundException;
+import java.util.HashMap;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.vertx.core.Future.failedFuture;
+import static java.lang.String.format;
+import static org.folio.HttpStatus.HTTP_NOT_FOUND;
+import static org.folio.HttpStatus.HTTP_OK;
+import static org.folio.verticle.consumers.util.QMEventTypes.QM_RECORD_UPDATED;
```
Comment (psmagin) [inline mod-source-record-manager-server/src/main/java/org/folio/services/ParsedRecordServiceImpl.java:30]: Your import is not in proper order.

* import static all other
* empty line
* import static org.folio.*
* empty line
* import java/javax.*
* empty line
* import all other
* empty line
* import org.folio.*

File: mod-source-record-manager-server/src/test/java/org/folio/rest/impl/mappingRulesProvider/MappingRulesProviderAPITest.java [hunk 5]
```diff
@@ -18,21 +19,49 @@
...
     Map defaultRules =
       RestAssured.given()
         .spec(spec)
         .when()
-        .get(SERVICE_PATH)
+        .get(SERVICE_PATH + MARC_BIB)
+        .then()
+        .statusCode(HttpStatus.SC_OK)
+        .extract().body().as(Map.class);
+    Assert.assertNotNull(defaultRules);
```
Comment (psmagin) [inline mod-source-record-manager-server/src/test/java/org/folio/rest/impl/mappingRulesProvider/MappingRulesProviderAPITest.java:37]: You should check that these rules are exactly rules for marc bib

File: mod-source-record-manager-server/src/test/java/org/folio/rest/impl/mappingRulesProvider/MappingRulesProviderAPITest.java [hunk 6]
```diff
@@ -18,21 +19,49 @@
...
+  public void shouldReturnDefaultMarcHoldingsRulesOnGet() {
+    Map defaultRules =
+      RestAssured.given()
+        .spec(spec)
+        .when()
+        .get(SERVICE_PATH + MARC_HOLDINGS)
         .then()
         .statusCode(HttpStatus.SC_OK)
         .extract().body().as(Map.class);
     Assert.assertNotNull(defaultRules);
```
Comment (psmagin) [inline mod-source-record-manager-server/src/test/java/org/folio/rest/impl/mappingRulesProvider/MappingRulesProviderAPITest.java:51]: You should check that these rules are exactly rules for marc holdings

File: ramls/mapping-rules-provider.raml [hunk 7]
```diff
@@ -23,22 +25,26 @@ resourceTypes:
...
-        description: "Not found"
-        body:
-          text/plain:
-            example: "Not found"
-      500:
-        description: "Internal server error"
-        body:
-          text/plain:
-            example: "Internal server error"
+  /{recordType}:
```
Comment (psmagin) [inline ramls/mapping-rules-provider.raml:28]: You should specify the type of `recordType` parameter

**Review by psmagin — CHANGES_REQUESTED** [review]
Add ticket number to PR title
Use ticket number in the commit messages and make them more meaningful, don't just copy ticket title.

File: mod-source-record-manager-server/src/main/java/org/folio/services/util/QueryPathUtil.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
+package org.folio.services.util;
+
+import javax.ws.rs.BadRequestException;
+
+import org.folio.Record;
+
+public class QueryPathUtil {
```
Comment (psmagin) [inline mod-source-record-manager-server/src/main/java/org/folio/services/util/QueryPathUtil.java:7]: Make class final

File: mod-source-record-manager-server/src/main/java/org/folio/services/util/QueryPathUtil.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+  private QueryPathUtil() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  /**
+   * Convert Sttong param to Folio RecordType enum
```
Comment (psmagin) [inline mod-source-record-manager-server/src/main/java/org/folio/services/util/QueryPathUtil.java:14]:
```suggestion
   * Convert string query param to {@link Record.RecordType}
```

File: mod-source-record-manager-server/src/main/java/org/folio/services/util/QueryPathUtil.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+   * @param recordType String param from URL path
+   * @return org.folio.Record.RecordType
+   */
+  public static Record.RecordType convert(String recordType) {
+    switch (recordType) {
+      case "marc-bib":
+        return Record.RecordType.MARC_BIB;
+      case "marc-holdings": return Record.RecordType.MARC_HOLDING;
+      default: throw new BadRequestException("Only marc-bib or marc-holdings supports");
+    }
```
Comment (psmagin) [inline mod-source-record-manager-server/src/main/java/org/folio/services/util/QueryPathUtil.java:25]: Create a pre-populated map in your class that contains the mapping between string and record type. 
For exception will be better to throw it in your controller or just without exception check for nullable.
```suggestion
    return Optional.ofNullable(queryToRecordTypeMap.get(recordType));
```
Also, method should have better naming, for example `toRecordType(String queryParam)`

File: mod-source-record-manager-server/pom.xml [hunk 11]
```diff
@@ -232,6 +232,17 @@
...
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <version>3.20.2</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>commons-lang</groupId>
+      <artifactId>commons-lang</artifactId>
+      <version>2.6</version>
+    </dependency>
```
Comment (psmagin) [inline mod-source-record-manager-server/pom.xml]: I'm pretty sure that this dependency already exists in the classpath.

File: mod-source-record-manager-server/pom.xml [hunk 12]
```diff
@@ -232,6 +232,17 @@
       <artifactId>util</artifactId>
       <version>33.0.3</version>
     </dependency>
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <version>3.20.2</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (psmagin) [inline mod-source-record-manager-server/pom.xml]: I don't see where you use this dependency.

Comment (Serhii-Zdorenko) [reply mod-source-record-manager-server/pom.xml]: image

It's not my changes...

File: mod-source-record-manager-server/pom.xml (hunk 11 above)
Comment (Serhii-Zdorenko) [reply mod-source-record-manager-server/pom.xml]: image
It's not my changes...

**Review by psmagin — APPROVED** [review]

File: mod-source-record-manager-server/src/main/resources/rules/marc_holding_rules.json [hunk 13]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (oleksandr-dekin) [inline mod-source-record-manager-server/src/main/resources/rules/marc_holding_rules.json:1]: It's not critical for back-end, but Khalilah mentioned, that we use mark holdings for the singular. So, in this case we should rename to `marc_holdings_rules.json`

File: mod-source-record-manager-server/src/main/resources/templates/db_scripts/schema.json [hunk 14]
```diff
@@ -162,6 +162,11 @@
       "run": "after",
       "snippetPath": "migrate_data_type_to_marc.sql",
       "fromModuleVersion": "mod-source-record-manager-3.2.0"
+    },
+    {
+      "run": "after",
+      "snippet": "CREATE TYPE rule_type AS ENUM ('MARC_BIB', 'MARC_HOLDING'); ALTER TABLE mapping_rules ADD COLUMN IF NOT EXISTS record_type rule_type NOT NULL DEFAULT 'MARC_BIB';",
```
Comment (oleksandr-dekin) [inline mod-source-record-manager-server/src/main/resources/templates/db_scripts/schema.json:168]: Please verify this script on the rancher

**Review by oleksandr-dekin — APPROVED** [review]

**Pull request merged**